### ISSUE TITLE
New buzzwords How Building the Perfect AI Pipeline Is Like Brewing the Perfect Shot of Espresso

### ISSUE THREAD

**Issue by postmillenial** [postmillenial/reading-list#1430]
How Building the Perfect AI Pipeline Is Like Brewing the Perfect Shot of Espresso<br>
http://ift.tt/2AsvO7n<br><img width="640" height="426" src="http://ift.tt/2zstpg2" class="webfeedsFeaturedVisual wp-post-image wp-stateless-item" style="display: block; margin: auto; margin-bottom: 20px;" data-image-size="large" data-stateless-media-bucket="cdn.thenewstack.io" data-stateless-media-name="media/2017/11/aa188e5e-coffe-1354786_1280.jpg"><a href="https://www.sigopt.com" class="clearfix infoBlock" target="_blank"></a>
<div class="infoBlockImageBlock"><img class="infoBlockImage" src="http://ift.tt/2zLQ6wr" style="border-radius: 50%;"></div>
<div class="infoBlockTextBlock">
<div class="infoBlockTitle">Scott Clark, co-founder and CEO of SigOpt</div>
<div class="infoBlockText">Scott Clark is the co-founder and CEO of SigOpt. He has been applying optimal learning techniques in industry and academia for years, from bioinformatics to production advertising systems. Before SigOpt, Scott worked on the ad targeting team at Yelp leading the charge on academic research and outreach with projects like the Yelp Dataset Challenge and open sourcing MOE. Scott holds a PhD in Applied Mathematics and an MS in Computer Science from Cornell University and BS degrees in Mathematics, Physics, and Computational Physics from Oregon State University. Scott was chosen as one of Forbes&rsquo; 30 under 30 in 2016.</div>
</div>
<p>Building the best AI pipeline is strikingly similar to crafting the perfect shot of espresso. In both cases, there are a multitude of tunable parameters that must be configured before the process even begins that have a huge impact on the end result.</p>
<p>From the water pressure and grind of an espresso bean to the learning rate and the number of hidden layers in a neural network, these configuration parameters can make or break your AI pipeline or perfect morning shot. There are many analogous components between these tunable systems, but in both cases getting to the best result has historically been more art than science. In this post, we&rsquo;ll discuss the similarities of both processes as well as a better way to tune them and similar complex systems.</p>
<p>Training a <a href="http://ift.tt/2xAZfDT" class="local-link">neural network</a> is not so different from brewing the perfect espresso (<a href="http://ift.tt/2zrukgO" target="_blank" data-saferedirecturl="http://ift.tt/2zLQ73t" class="ext-link" rel="external nofollow">or baking the perfect batch of cookies</a>). Water passes through a network of ground espresso beans at some temperature and pressure, absorbing flavor and caffeine as it subtly transforms the beans and eventually outputs a delicious shot of espresso on the other side. As data passes through a deep learning pipeline it transforms the weights of its neurons, absorbing information and converging to a model that can be applied to tasks as varied as <a href="http://ift.tt/2oPgpwk" target="_blank" data-saferedirecturl="http://ift.tt/2zNcF3D" class="ext-link" rel="external nofollow">natural language processing</a> or <a href="http://ift.tt/2vwasrr" target="_blank" data-saferedirecturl="http://ift.tt/2zNcG7H" class="ext-link" rel="external nofollow">sequence classification</a>.</p>
<p>The water that passes through the espresso beans is similar to the data coursing through a neural network as it is trained. It passes through the system to become something different and desirable at the end, by interacting with and changing the properties of the system itself as it goes. One can tune various aspects of how the water or data interact with the system like pressure and temperature to affect how it is absorbed by the espresso beans or the learning rate and <a href="http://ift.tt/1rIE5iQ" class="ext-link" rel="external nofollow">Stochastic Gradient Descent</a> (SGD) parameters in a neural network. These tunable parameters will affect how the network of coffee grounds or neurons are influenced by the water or data.</p>
<div id="attachment_3457621" style="width: 579px;" class="wp-caption aligncenter">
<img class="size-full wp-image-3457621" src="http://ift.tt/2zLQ87x" width="569" height="557" data-id="3457621"><p class="wp-caption-text">Water pressure and temperature affect how much flavor and caffeine are added as the water passes through the ground espresso beans.</p>
</div>
<p>&nbsp;</p>
<div class="mceTemp"></div>
<div style="width: 526px;" class="wp-caption aligncenter">
<img src="http://ift.tt/2zstpN4" width="516" height="172"><p class="wp-caption-text">Stochastic Gradient Descent (SGD) parameters affect how each neuron in the deep learning system converges to the optimal weight as the data passes through the neural network.</p>
</div>
<p>Furthermore, the espresso beans themselves are similar to the overall architecture of a deep learning pipeline. One variation of espresso bean versus another can have the same difference on the shot as the decision between a convolutional or recurrent neural network in a deep learning pipeline. The grind of the beans and amount used define the lattice over which the water passes, like the architecture parameters in a neural network of hidden layers and number of neurons per layer.</p>
<p>Roasting the beans in different ways will change how the flavor is absorbed by each ground particle of espresso bean, like the activation function in a neural network allows data to have different effects on individual neurons within the network. Different espresso machines have an impact on the shot even with otherwise identical configuration parameters just like deep learning frameworks like <a href="http://ift.tt/2ikr0Na" target="_blank" data-saferedirecturl="http://ift.tt/2zrwAEV" class="ext-link" rel="external nofollow">MXNet</a>, <a href="http://ift.tt/1Qp9srs" target="_blank" data-saferedirecturl="http://ift.tt/2zstqk6" class="ext-link" rel="external nofollow">TensorFlow</a>, and <a href="https://caffe2.ai/" target="_blank" data-saferedirecturl="http://ift.tt/2zNA5Gk" class="ext-link" rel="external nofollow">Caffe2</a> can impart subtle differences onto the trained model.</p>
<p>Time in both systems is analogous, the longer you pump water through the machine the more effect they will have on the eventual shot (in both size and quality), similarly the longer you train a neural network (the number of epochs you train over) will have an effect on what is learned, how much it converges, and what the results are.</p>
<div id="attachment_3457620" style="width: 650px;" class="wp-caption aligncenter">
<img class="wp-image-3457620 size-full" src="http://ift.tt/2zrwBbX" width="640" height="426" data-id="3457620"><p class="wp-caption-text">The amount of espresso beans and the coarseness of the grind change how the water picks up distinct amounts of caffeine and flavor.</p>
</div>
<div id="attachment_3457619" style="width: 998px;" class="wp-caption aligncenter">
<img class="wp-image-3457619 size-full" src="http://ift.tt/2zLQafF" width="988" height="550" data-id="3457619"><p class="wp-caption-text">The architecture of the neural network, like the number of hidden layers and number of neurons per layer change what is learned from the data.</p>
</div>
<p>Once you&rsquo;ve settled on a configuration of parameters to try and brewed your espresso or trained your network, the next step is measuring the output to determine whether or not it was a success. The best shot of an espresso makes the perfect tradeoff between flavor and caffeine, the bitterness needs to be just right. Similarly, a deep learning pipeline has specific, sometimes competing metrics, around accuracy, robustness, or speed. Furthermore, the results of brewing an espresso can be post-processed into a latte, cappuccino, or americano, each with their own tunable parameters in the same way that a neural network output can be fed into a larger pipeline for applications like fraud detection or algorithmic trading. Finding the best configuration for these systems as efficiently as possible can result in boosted performance for these models or a better caffeine boost in the morning. All of the parameters interact to influence the desired output in different, often non-intuitive ways.</p>
<blockquote>
<p>The black box nature of these methods allows us to consider the espresso brewing problem identically to tuning an AI pipeline.</p>
</blockquote>
<p>The art of creating the best espresso has been honed by masters since the 19th century. Similarly, deep learning pipelines are often tuned in practice via expert intuition, especially when popular brute force approaches like <a href="http://ift.tt/2zrwCg1" target="_blank" data-saferedirecturl="http://ift.tt/2zM5sks" class="ext-link" rel="external nofollow">grid and random search</a> become intractable due to the sheer quantity of configuration parameters or expense of training the model. Unfortunately, the optimal configuration of these pipelines for different applications can be wildly different on different datasets and in different contexts. So the manual tuning approach often boils down to time-consuming and expensive trial-and-error optimization, often in high dimensions, which wastes precious expert and computational resources.</p>
<p dir="ltr">Bayesian optimization techniques allow for finding the best configurations of these systems in as few attempts as possible. This is done by drawing from academic research in fields like <a href="http://ift.tt/1sZcuYR" class="ext-link" rel="external nofollow">Optimal Learning</a> and <a href="http://ift.tt/2zN0Ieu" class="ext-link" rel="external nofollow">Sequential Model-Based Optimization</a>. These techniques trade-off exploration, learning more about how the parameters interact and combine to influence the desired result, and exploitation, using what we already know to drive towards better performance.</p>
<p dir="ltr">The black box nature of these methods allows us to consider the espresso brewing problem identically to tuning an AI pipeline. Black box methods only observe the input to a system (the specific configuration to be evaluated) and the output (the desired objective or set of objectives to be optimized). This allows Bayesian Optimization tools to easily bolt on top of any underlying system without requiring any information about proprietary data or models. In fact, companies like <a href="http://ift.tt/2bGymG1" target="_blank" data-saferedirecturl="http://ift.tt/2zMFyNp" class="ext-link" rel="external nofollow">MillerCoors have applied this to another type of brewing</a>.</p>
<p>By getting to the best configurations exponentially faster than common techniques like grid search, and not relying on humans to use intuition to do high dimensional optimization in their head, Bayesian optimization allows you to get to the best version of your model, or perfect espresso shot, faster and cheaper than ever before.</p>
<p class="attribution" dir="ltr">Feature image <a href="http://ift.tt/2zstroa" class="ext-link" rel="external nofollow">via</a> Pixabay.</p>
<p>The post <a rel="nofollow" href="http://ift.tt/2AsvO7n">How Building the Perfect AI Pipeline Is Like Brewing the Perfect Shot of Espresso</a> appeared first on <a rel="nofollow" href="https://thenewstack.io">The New Stack</a>.</p>
<br><br>
via Blog &ndash; The New Stack https://thenewstack.io<br>
November 10, 2017 at 12:11PM
